File: matlab/stratified_resample.m
function [keep, Neff] = stratified_resample(w)
%function [keep, Neff] = stratified_resample(w)
%
% INPUT:
%   w - set of N weights [w1, w2, ..]
%
% OUTPUTS:
%   keep - N indices of particles to keep 
%   Neff - number of effective particles (measure of weight variance)
%
% Tim Bailey 2004.


w= w / sum(w); % normalise
Neff= 1 / sum(w .^ 2); 

len= length(w);
keep= zeros(1,len);
select = stratified_random(len); 
w= cumsum(w); 

ctr=1; 
for i=1:len
   while ctr<=len & select(ctr)<w(i)
       keep(ctr)= i;
       ctr=ctr+1; 
   end
end

File: cpp/stratified_resample.cpp
#include "stratified_resample.h"
#include "stratified_random.h"
#include <iostream>

using namespace std;

void stratified_resample(VectorXf w, vector<int> &keep, float &Neff)
{
    VectorXf wsqrd(w.size());
    for (int i=0; i<w.size(); i++) {
        w(i) = w(i)/w.sum();
        wsqrd(i) = pow(w(i),2);
    }
    Neff = 1/wsqrd.sum();

    int len = w.size();
    keep.resize(len);
    for (int i=0; i<len; i++) {
        keep[i] = -1;
    }

    vector<float> select;
    stratified_random(len,select); 
    cumsum(w);    

    int ctr=0;
    for (int i=0; i<len; i++) {
        while ((ctr<len) && (select[ctr]<w(i))) {
            keep[ctr] = i;
            ctr++;
        }
    }
}

//
//returns a cumulative sum array
//
void cumsum(VectorXf &w) 
{
    VectorXf csumVec(w.size());
    for (int i=0; i< w.size(); i++) {
        float sum =0;
        for (int j=0; j<=i; j++) {
            sum+=w(j);
        }			
        csumVec(i) = sum;
    }
    w = VectorXf(csumVec); //copy constructor. Double check
}
